pages/home/build-tools/create-a-tool-with-secrets.mdx

```diff
@@ -0,0 +1,216 @@
+    return tables
+
+
+@tool(requires_secrets=["DATABASE_CONNECTION_STRING"])
+def get_table_schema(context: ToolContext, schema_name: Annotated[str, "The database schema to get the table schema of"], table_name: Annotated[str, "The table to get the schema of"]) -> list[str]:
+    """Get the schema of a table"""
+    engine = _get_engine(context.get_secret("DATABASE_CONNECTION_STRING"))
+    return _get_table_schema(engine, schema_name, table_name)
+
+
+@tool(requires_secrets=["DATABASE_CONNECTION_STRING"])
+def execute_query(context: ToolContext, query: Annotated[str, "The SQL query to execute"]) -> list[str]:
+    """Execute a query and return the results"""
+    engine = _get_engine(context.get_secret("DATABASE_CONNECTION_STRING"))
+    return _execute_query(engine, query)
+
+
+def _get_engine(connection_string: str) -> Engine:
+    """Get a connection to the database.  Note that we build the engine with an isolation level of READ UNCOMMITTED to prevent all writes."""
+    return create_engine(connection_string, isolation_level='READ UNCOMMITTED')
+
+
+def _get_tables(engine: Engine, schema_name: str) -> list[str]:
+    """Get all the tables in the database"""
+    inspector = inspect(engine)
+    schemas = inspector.get_schema_names()
+    tables = []
+    for schema in schemas:
+        if schema == schema_name:
+            tables.extend(inspector.get_table_names(schema=schema))
+    return tables
+
+def _get_table_schema(engine: Engine, schema_name: str, table_name: str) -> list[str]:
+    """Get the schema of a table"""
+    inspector = inspect(engine)
+    columns_table = inspector.get_columns(table_name, schema_name)
+    return [f"{column['name']}: {column['type'].python_type.__name__}" for column in columns_table]
+
+def _execute_query(engine: Engine, query: str) -> list[str]:
+    """Execute a query and return the results."""
+    with engine.connect() as connection:
+        result = connection.execute(text(query))
+        return [str(row) for row in result.fetchall()]
+
+```
+
+### Use your tool with Arcade
+
+Now you can use your custom authorized tool with Arcade in your application.
+
+Here's an example of how to use your tool.  Note that for this example, the table schema includes a `users` table and a `messages` table, and `messsages` has a foreign key back to `users`.
+
+<details>
+
+<summary>See full schema</summary>
+
+```sql
+CREATE TABLE "public"."users" (
+    "id" serial PRIMARY KEY NOT NULL,
+	"name" varchar(256) NOT NULL,
+	"email" text NOT NULL UNIQUE,
+	"password_hash" text NOT NULL,
+	"created_at" timestamp DEFAULT now() NOT NULL,
+	"updated_at" timestamp DEFAULT now() NOT NULL
+);
+
+CREATE TABLE "public"."messages" (
+    "id" serial PRIMARY KEY NOT NULL,
+	"body" text NOT NULL,
+	"user_id" integer NOT NULL REFERENCES "public"."users" (id),
+	"created_at" timestamp DEFAULT now() NOT NULL,
+	"updated_at" timestamp DEFAULT now() NOT NULL
+);
+```
+</details>
+
+
+```ts
```

Comment:
when possible, utilizing [remark-code-import](https://github.com/ArcadeAI/docs/blob/031d07581ec069560f8362cc61d2f0afe602c332/package.json#L47) is preferred. See [this file](https://github.com/ArcadeAI/docs/blob/031d07581ec069560f8362cc61d2f0afe602c332/pages/home/build-tools/retry-tools-with-improved-prompt.mdx?plain=1#L26) on how to use remark-code-import. The goal is to have all code in their own files (there's some debt to move lingering code snippets into their own files across the documentation repo)